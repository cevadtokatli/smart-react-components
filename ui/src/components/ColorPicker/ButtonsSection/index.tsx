import Section from '@smart-react-components/core/Element/Section'
import { Theme } from '@smart-react-components/core/theme'
import { PaletteProp } from '@smart-react-components/core/types'
import React from 'react'
import { ThemeContext } from 'styled-components'
import Button from '../../../Button'
import ButtonList from '../../../Button/ButtonList'
import ColorPickerProps from '../../../context/ColorPickerProps'
import SaveIcon from '../../../icons/Save'
import CancelIcon from '../../../icons/Cancel'
import { applyResponsiveStyledProp } from '../../../util/props'

interface Props {
  isSoft: boolean
  cancelLabel?: string
  saveLabel?: string
  onCancel: (e: Event) => void
  onSave: (e: Event) => void
  palette: PaletteProp
}

const ButtonsSection: React.FC<Props> = ({ cancelLabel, saveLabel, isSoft, onCancel, onSave, palette }) => {
  const { colorPickerSize, colorPickerSizeSm, colorPickerSizeMd, colorPickerSizeLg, colorPickerSizeXl, isDisabled } = React.useContext(ColorPickerProps)
  const theme = React.useContext<Theme>(ThemeContext)

  return (
    <Section>
      <ButtonList
        hasSpace={false}
        isBlock
        isSoft={isSoft}
        palette={palette}
        size={colorPickerSize}
        sizeSm={colorPickerSizeSm}
        sizeMd={colorPickerSizeMd}
        sizeLg={colorPickerSizeLg}
        sizeXl={colorPickerSizeXl}
      >
        <Button
          {...applyResponsiveStyledProp({ colorPickerSize, colorPickerSizeSm, colorPickerSizeMd, colorPickerSizeLg, colorPickerSizeXl }, 'colorPickerSize', 'marginRight', v => `$size.colorPicker.${v}.space`)}
          isDisabled={isDisabled}
          onClick={e => onSave(e as any)}
        >
          <SaveIcon />
          <span>{saveLabel ?? theme.$.i18n.save}</span>
        </Button>
        <Button
          isDisabled={isDisabled}
          onClick={e => onCancel(e as any)}
        >
          <CancelIcon />
          <span>{cancelLabel ?? theme.$.i18n.cancel}</span>
        </Button>
      </ButtonList>
    </Section>
  )
}

export default ButtonsSection
