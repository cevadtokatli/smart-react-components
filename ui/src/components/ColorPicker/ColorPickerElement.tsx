import Div from '@smart-react-components/core/Element/Div'
import { StyledProps } from '@smart-react-components/core/styled-props'
import { ResponsiveProp, SizeProp } from '@smart-react-components/core/types'
import styled from 'styled-components'

interface Props extends
  StyledProps,
  ResponsiveProp<'colorPickerSize', SizeProp> {
  isOutline: boolean
}

export default styled(Div).attrs<Props>({
  getColorPickerSize: (v, t) => `
    padding: ${t.$.size.colorPicker[v].space};
    width: ${t.$.size.colorPicker[v].width};

    > *:not(:first-child) {
      padding-top: ${t.$.size.colorPicker[v].space};
    }
  `,
})<Props>(({ theme, isOutline }: Props) => `
  border-radius: ${theme.$.radius.colorPicker};
  box-sizing: border-box;

  ${!isOutline
    ? `
      background: ${theme.$.color.dynamic.gray};
    `
    : ''
  }
`)
