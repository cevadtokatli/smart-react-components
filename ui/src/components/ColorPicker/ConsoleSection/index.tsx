import { Color, Nullable, PaletteProp } from '@smart-react-components/core/types'
import React from 'react'
import ColorPickerProps from '../../../context/ColorPickerProps'
import Eraser from '../../../icons/Eraser'
import Palette from '../../../icons/Palette'
import Input from '../../../Input'
import InputAddon from '../../../Input/InputAddon'
import { ColorPickerFormat, ConsoleRef } from '../../../types/color-picker'
import { colorToString, isLight } from '../../../util/color-picker'
import ConsoleSectionElement from './ConsoleSectionElement'

interface Props {
  format: ColorPickerFormat
  initialValue: Nullable<Color>
  palette: PaletteProp
  onRef: (v: ConsoleRef) => void
}

const ConsoleSection: React.FC<Props> = ({ format, initialValue, onRef, palette }) => {
  const { colorPickerSize, colorPickerSizeSm, colorPickerSizeMd, colorPickerSizeLg, colorPickerSizeXl, isDisabled, updateValue, value } = React.useContext(ColorPickerProps)

  const clearButtonEl = React.useRef<HTMLDivElement>(null)
  const initialButtonEl = React.useRef<HTMLDivElement>(null)
  const inputEl = React.useRef<HTMLInputElement>(null)

  const handleChange = (value: Nullable<Color>) => {
    const background = value?.toString() ?? 'transparent'
    const isDark = String(value ? isLight(value, true) : true)
    clearButtonEl.current.style.background = background
    clearButtonEl.current.setAttribute('data-src-is-dark', isDark)
    inputEl.current.style.background = background
    inputEl.current.setAttribute('data-src-is-dark', isDark)
    inputEl.current.value = colorToString(value, format)
  }

  const handleDrag = (value: Nullable<Color>) => {
    handleChange(value)
  }

  const handleInputKeyDown = (e: React.KeyboardEvent<HTMLElement>) => {
    if (e.key === 'Enter') {
      updateValue((e.target as HTMLInputElement).value)
    }
  }

  React.useEffect(() => {
    onRef({
      onDrag: handleDrag,
    })
  }, [])

  React.useEffect(() => {
    handleChange(value)
  }, [value])

  return (
    <ConsoleSectionElement
      isDisabled={isDisabled}
      palette={palette}
    >
      <Input
        hasBorder={false}
        leftAddon={(
          <InputAddon
            elementProps={{
              'data-src-is-dark': initialValue ? String(isLight(initialValue, true)) : true,
              style: {
                background: initialValue?.toString() ?? 'transparent',
              },
            }}
            isSeparated
            ref={initialButtonEl}
            {...(!isDisabled && { onClick: () => updateValue(initialValue) })}
          >
            <Palette
            />
          </InputAddon>
        )}
        ref={inputEl}
        rightAddon={(
          <InputAddon
            isSeparated
            ref={clearButtonEl}
            {...(!isDisabled && { onClick: () => updateValue(null) })}
          >
            <Eraser />
          </InputAddon>
        )}
        size={colorPickerSize}
        sizeSm={colorPickerSizeSm}
        sizeMd={colorPickerSizeMd}
        sizeLg={colorPickerSizeLg}
        sizeXl={colorPickerSizeXl}
        {...!isDisabled && {
          onBlur: e => updateValue((e.target as HTMLInputElement).value),
          onKeyDown: handleInputKeyDown,
        }}
      />
    </ConsoleSectionElement>
  )
}

export default ConsoleSection
