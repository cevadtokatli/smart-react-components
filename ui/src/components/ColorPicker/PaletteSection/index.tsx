import Section from '@smart-react-components/core/Element/Section'
import { SetState } from '@smart-react-components/core/types'
import CSSTransition from '@smart-react-components/transition/CSSTransition'
import React from 'react'
import { ColorPickerFormat } from '../../../ColorPicker'
import { defaultPaletteColors } from '../../../constants/color-picker'
import ColorPickerProps from '../../../context/ColorPickerProps'
import ArrowMultipleDownIcon from '../../../icons/ArrowMultipleDown'
import { applyResponsiveStyledProp } from '../../../util/props'
import PaletteList from './PaletteList'

interface Props {
  canAddColorToPalette: boolean
  format: ColorPickerFormat
  paletteColors?: string[]
  setPaletteColors?: SetState<string[]>
}

const PaletteSection: React.FC<Props> = ({ canAddColorToPalette, format, paletteColors, setPaletteColors }) => {
  const { colorPickerSize, colorPickerSizeSm, colorPickerSizeMd, colorPickerSizeLg, colorPickerSizeXl } = React.useContext(ColorPickerProps)
  const [localPaletteColors, setLocalPaletteColors] = React.useState<string[]>(() => defaultPaletteColors)
  const [status, setStatus] = React.useState<boolean>(() => false)

  return (
    <>
      <Section>
        <ArrowMultipleDownIcon
          {...applyResponsiveStyledProp({ colorPickerSize, colorPickerSizeSm, colorPickerSizeMd, colorPickerSizeLg, colorPickerSizeXl }, 'colorPickerSize', 'iconSize', v => `$size.icon.${v}`)}
          display="block"
          fill="$color.gray500"
          margin="auto"
          onClick={() => setStatus(!status)}
        />
      </Section>
      <CSSTransition className="src-fade" duration={150} status={status}>
        <PaletteList
          canAddColorToPalette={canAddColorToPalette}
          format={format}
          paletteColors={paletteColors ?? localPaletteColors}
          setPaletteColors={setPaletteColors ?? setLocalPaletteColors}
        />
      </CSSTransition>
    </>
  )
}

export default PaletteSection
