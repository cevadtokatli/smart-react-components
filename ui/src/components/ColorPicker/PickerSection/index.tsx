import Section from '@smart-react-components/core/Element/Section'
import { Color, Nullable } from '@smart-react-components/core/types'
import { addEventListener, removeEventListener } from '@smart-react-components/core/util/dom'
import React from 'react'
import ColorPickerProps from '../../../context/ColorPickerProps'
import { PickerRef } from '../../../types/color-picker'
import { isEqual } from '../../../util/color-picker'
import MajorPicker from './MajorPicker'
import MinorPicker from './MinorPicker'
import OpacityPicker from './OpacityPicker'

interface Props {
  hasOpacityPicker: boolean
  onDrag: (value: Color) => void
}

const PickerSection: React.FC<Props> = ({ hasOpacityPicker, onDrag }) => {
  const { updateValue, value } = React.useContext(ColorPickerProps)
  const color = React.useRef<Nullable<Color>>(undefined)
  const majorPicker = React.useRef<PickerRef>(null)
  const minorPicker = React.useRef<PickerRef>(null)
  const opacityPicker = React.useRef<PickerRef>(null)

  const handleStart = (onMove: (e: Event) => void, onEnd: () => void) => {
    addEventListener(document, ['mousemove', 'touchmove'], onMove)
    addEventListener(document, ['mouseup', 'touchend'], onEnd)
    document.body.style.userSelect = 'none'
  }

  const handleEnd = (onMove: (e: Event) => void, onEnd: () => void) => {
    removeEventListener(document, ['mousemove', 'touchmove'], onMove)
    removeEventListener(document, ['mouseup', 'touchend'], onEnd)
    document.body.style.userSelect = 'initial'
    updateValue(color.current)
  }

  const handleMajorPickerDrag = () => {
    opacityPicker.current?.onExternalDrag()
    handleDrag()
  }

  const handleMinorPickerDrag = () => {
    majorPicker.current.onExternalDrag()
  }

  const handleDrag = () => onDrag(color.current)

  React.useEffect(() => {
    if (typeof color.current === 'undefined' || !isEqual(color.current, value)) {
      color.current = value
      majorPicker.current.onChange()
      minorPicker.current.onChange()
      opacityPicker.current.onChange()
    }
  }, [value])

  return (
    <Section display="flex" justifyContent="space-between">
      <MajorPicker
        color={color}
        onEnd={handleEnd}
        onDrag={handleMajorPickerDrag}
        onRef={v => { majorPicker.current = v } }
        onStart={handleStart}
      />
      <MinorPicker
        color={color}
        onEnd={handleEnd}
        onDrag={handleMinorPickerDrag}
        onRef={v => { minorPicker.current = v } }
        onStart={handleStart}
      />
      { hasOpacityPicker && (
        <OpacityPicker
          color={color}
          onEnd={handleEnd}
          onDrag={handleDrag}
          onRef={v => { opacityPicker.current = v } }
          onStart={handleStart}
        />
      )}
    </Section>
  )
}

export default PickerSection
