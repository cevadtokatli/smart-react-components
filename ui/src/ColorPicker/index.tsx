import { Hr } from '@smart-react-components/core'
import extractElementProps from '@smart-react-components/core/element-props'
import intrinsicStyledCoreProps, { IntrinsicStyledCoreProps } from '@smart-react-components/core/element-props/intrinsic-styled-core-props'
import useChangeEffect from '@smart-react-components/core/hooks/useChangeEffect'
import { Color, Nullable, PaletteProp, Partial, ResponsiveProp, SetState, SizeProp } from '@smart-react-components/core/types'
import { getColor } from '@smart-react-components/core/util/color'
import React from 'react'
import ButtonsSection from '../components/ColorPicker/ButtonsSection'
import ColorPickerElement from '../components/ColorPicker/ColorPickerElement'
import ConsoleSection from '../components/ColorPicker/ConsoleSection'
import PaletteSection from '../components/ColorPicker/PaletteSection'
import PickerSection from '../components/ColorPicker/PickerSection'
import ColorPickerProps from '../context/ColorPickerProps'
import { ColorPickerFormat } from '../types/color-picker'
import { colorToString } from '../util/color-picker'

export { ColorPickerFormat } from '../types/color-picker'

export interface Props extends
  IntrinsicStyledCoreProps,
  Partial<ResponsiveProp<'size', SizeProp>> {
  canAddColorToPalette?: boolean
  cancelLabel?: string
  format?: ColorPickerFormat
  hasButtons?: boolean
  hasConsole?: boolean
  hasPalette?: boolean
  isDisabled?: boolean
  isOutline?: boolean
  isSoft?: boolean
  hasOpacityPicker?: boolean
  palette?: PaletteProp
  paletteColors?: string[]
  onCancel?: (e: Event) => void
  onDrag?: (e: Event) => void
  onSave?: (e: Event) => void
  saveLabel?: string
  setPaletteColors?: SetState<string[]>
  setValue: SetState<string>
  value: string
}

const ColorPicker: React.FC<Props> = props => {
  const consoleSection = React.useRef<{
    onDrag: (value: Color) => void
  }>(null)

  const [value, setValue] = React.useState<Nullable<Color>>(() => props.value ? getColor(props.value) : null)
  const [initialValue, setInitialValue] = React.useState<Nullable<Color>>(() => props.value ? getColor(props.value) : null)

  const updateValue = (value: Nullable<string | Color>) => props.setValue(value !== null
    ? colorToString(typeof value === 'string' ? getColor(value) : value, props.format)
    : null,
  )

  const handleDrag = (value: Color) => {
    props.onDrag?.(value)

    consoleSection.current?.onDrag(value)
  }

  const handleCancel = (e: Event) => {
    props.onCancel?.(e)

    if (!e.defaultPrevented) {
      updateValue(initialValue)
    }
  }

  const handleSave = (e: Event) => {
    props.onSave?.(e)

    if (!e.defaultPrevented) {
      setInitialValue(value)
    }
  }

  useChangeEffect(() => {
    const newValue = props.value ? getColor(props.value) : null
    if (newValue !== value) {
      setValue(newValue)
    }
  }, [props.value])

  return (
    <ColorPickerProps.Provider
      value={{
        colorPickerSize: props.size,
        colorPickerSizeSm: props.sizeSm,
        colorPickerSizeMd: props.sizeMd,
        colorPickerSizeLg: props.sizeLg,
        colorPickerSizeXl: props.sizeXl,
        isDisabled: props.isDisabled,
        palette: props.palette,
        updateValue,
        value,
      }}
    >
      <ColorPickerElement
        {...extractElementProps(props, [intrinsicStyledCoreProps])}
        colorPickerSize={props.size}
        colorPickerSizeSm={props.sizeSm}
        colorPickerSizeMd={props.sizeMd}
        colorPickerSizeLg={props.sizeLg}
        colorPickerSizeXl={props.sizeXl}
        isOutline={props.isOutline}
        palette={props.palette}
      >
        <PickerSection
          hasOpacityPicker={props.hasOpacityPicker}
          onDrag={handleDrag}
        />
        { (props.hasConsole || props.hasButtons) && (
          <Hr
            border="none"
            background="transparent"
            width="100%"
          />
        ) }
        { props.hasConsole && (
          <ConsoleSection
            format={props.format}
            initialValue={initialValue}
            onRef={v => { consoleSection.current = v }}
            palette={props.palette}
          />
        ) }
        { props.hasButtons && (
          <ButtonsSection
            cancelLabel={props.cancelLabel}
            isSoft={props.isSoft}
            onCancel={handleCancel}
            onSave={handleSave}
            palette={props.palette}
            saveLabel={props.saveLabel}
          />
        ) }
        { props.hasPalette && (
          <PaletteSection
            canAddColorToPalette={props.canAddColorToPalette}
            format={props.format}
            palette={props.palette}
            paletteColors={props.paletteColors}
            setPaletteColors={props.setPaletteColors}
          />
        ) }
      </ColorPickerElement>
    </ColorPickerProps.Provider>
  )
}

ColorPicker.defaultProps = {
  format: ColorPickerFormat.HEX,
  canAddColorToPalette: true,
  hasButtons: true,
  hasConsole: true,
  hasOpacityPicker: true,
  hasPalette: true,
  size: 'medium',
  palette: 'primary',
}

export default ColorPicker
