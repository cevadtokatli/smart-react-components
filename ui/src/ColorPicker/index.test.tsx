import { wrapTheme } from '@smart-react-components/core/test'
import { render } from '@testing-library/react'
import React from 'react'
import ColorPicker from './'

describe('<ColorPicker />', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(wrapTheme(<ColorPicker value="red" setValue={jest.fn()} />))
    expect(asFragment()).toMatchSnapshot()
  })
})
