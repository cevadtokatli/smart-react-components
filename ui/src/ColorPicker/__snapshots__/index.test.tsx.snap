// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ColorPicker /> should match snapshot 1`] = `
<DocumentFragment>
  .c30 {
  margin-right: 10px;
  isolation: isolate;
  overflow: hidden;
  position: relative;
  -webkit-tap-highlight-color: transparent;
}

.c30 .src-button-content {
  font-size: .875rem;
  padding: .45rem .9rem;
}

.c30 .src-button-content > .c23 {
  height: 1.1rem;
  width: 1.1rem;
}

.c30 .src-button-content > :not(:last-child) {
  margin-right: 0.225rem;
}

.c33 {
  isolation: isolate;
  overflow: hidden;
  position: relative;
  -webkit-tap-highlight-color: transparent;
}

.c33 .src-button-content {
  font-size: .875rem;
  padding: .45rem .9rem;
}

.c33 .src-button-content > .c23 {
  height: 1.1rem;
  width: 1.1rem;
}

.c33 .src-button-content > :not(:last-child) {
  margin-right: 0.225rem;
}

.c0 {
  padding: 10px;
  width: 275px;
}

.c0 > *:not(:first-child) {
  margin-top: 10px;
}

.c3 {
  height: 175px;
  width: 175px;
}

.c5 {
  background: linear-gradient(to right,rgba(255,255,255,1),rgba(255,255,255,0));
}

.c6 {
  background: linear-gradient(to top,rgba(0,0,0,1),rgba(0,0,0,0));
  box-shadow: 0 0 3px rgba(0,0,0,.3) inset;
}

.c10 {
  height: 175px;
  width: 30px;
}

.c13 {
  background: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAwAAAAMCAIAAADZF8uwAAAAGUlEQVQYV2M4gwH+YwCGIasIUwhT25BVBADtzYNYrHvv4gAAAABJRU5ErkJggg==');
}

.c14 {
  box-shadow: 0 0 3px rgba(0,0,0,.3) inset;
}

.c19 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
}

.c21 {
  font-size: 1rem;
  padding: 0.45rem;
}

.c21 .c23 {
  height: 1rem;
  width: 1rem;
}

.c15 {
  border: none;
  background: transparent;
  width: 100%;
}

.c26 {
  font-size: .875rem;
  padding: .45rem .9rem;
  padding-left: 0.45rem;
  padding-right: 0.45rem;
}

.c17 {
  display: block;
}

.c17 > span {
  margin-bottom: .5rem;
}

.c17 > span {
  font-size: 1rem;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c7 {
  height: 18px;
  width: 18px;
}

.c12 {
  height: 14px;
  width: 14px;
}

.c34 {
  height: 1rem;
  width: 1rem;
  display: block;
  fill: #b1bbc4;
  margin: auto;
}

.c31 {
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  background: transparent;
  border: 0;
  border-radius: .125rem;
  cursor: pointer;
  font-family: inherit;
  position: relative;
  -webkit-transition: box-shadow 300ms 0s ease-in-out;
  transition: box-shadow 300ms 0s ease-in-out;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  width: 100%;
}

.c31:active {
  outline: 0;
}

.c31 .src-button-content {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c31 .src-button-content {
  color: #fff;
  background: #0665d0;
  border-color: #0665d0;
}

.c31:hover .src-button-content {
  background: #055BBB;
  border-color: #055BBB;
}

.c31:hover .src-button-addon {
  background: #0552A9;
  border-color: #0552A9;
}

.c31:focus {
  box-shadow: 0 0 0 .15rem rgba(6,101,208,0.5);
}

.c31 .src-button-content {
  border-radius: .125rem;
}

.c32 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: solid 1px;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  fill: currentcolor;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  -webkit-transition: 300ms 0s ease-in-out;
  transition: 300ms 0s ease-in-out;
  -webkit-transition-property: background,border,color,fill;
  transition-property: background,border,color,fill;
  word-break: break-word;
  word-wrap: break-word;
}

.c29 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c29 > * {
  -webkit-flex: 1 1 50%;
  -ms-flex: 1 1 50%;
  flex: 1 1 50%;
  width: 50%;
}

.c1 {
  border-radius: .25rem;
  box-sizing: border-box;
  background: #eee;
}

.c18 > span {
  display: block;
  word-break: break-word;
  word-wrap: break-word;
  -webkit-tap-highlight-color: transparent;
  cursor: default;
}

.c27 {
  border: 0;
  border-radius: 0;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  font-family: inherit;
  min-width: 1px;
  -webkit-transition: 200ms 0s ease-in-out;
  transition: 200ms 0s ease-in-out;
  -webkit-transition-property: background,color,fill;
  transition-property: background,color,fill;
  -webkit-appearance: none;
  -webkit-tap-highlight-color: transparent;
  background: #fff;
  color: #323a46;
}

.c27::-webkit-input-placeholder {
  opacity: 1;
  color: #dbdbdb;
}

.c27::-moz-placeholder {
  opacity: 1;
  color: #dbdbdb;
}

.c27:-moz-placeholder {
  opacity: 1;
  color: #dbdbdb;
}

.c27:-ms-input-placeholder {
  opacity: 1;
  color: #dbdbdb;
}

.c27::-webkit-input-placeholder {
  opacity: 1;
  color: #dbdbdb;
}

.c27::-moz-placeholder {
  opacity: 1;
  color: #dbdbdb;
}

.c27:-ms-input-placeholder {
  opacity: 1;
  color: #dbdbdb;
}

.c27::placeholder {
  opacity: 1;
  color: #dbdbdb;
}

.c27:active,
.c27:focus {
  outline: 0;
}

.c24 {
  border: 0;
  border-radius: .2rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  min-width: 1px;
  overflow: hidden;
  -webkit-transition: 200ms 0s ease-in-out;
  transition: 200ms 0s ease-in-out;
  -webkit-transition-property: border;
  transition-property: border;
  border-color: #dbdbdb;
  border-left: 0;
  border-bottom-left-radius: 0;
  border-top-left-radius: 0;
  border-right: 0;
  border-bottom-right-radius: 0;
  border-top-right-radius: 0;
}

.c22 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  fill: currentcolor;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: 200ms 0s ease-in-out;
  transition: 200ms 0s ease-in-out;
  -webkit-transition-property: background,border,color,fill;
  transition-property: background,border,color,fill;
  border-bottom-left-radius: .2rem;
  border-top-left-radius: .2rem;
  background: #fff;
  border-color: #dbdbdb;
  color: #323a46;
  cursor: pointer;
}

.c28 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  fill: currentcolor;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: 200ms 0s ease-in-out;
  transition: 200ms 0s ease-in-out;
  -webkit-transition-property: background,border,color,fill;
  transition-property: background,border,color,fill;
  border-bottom-right-radius: .2rem;
  border-top-right-radius: .2rem;
  background: #fff;
  border-color: #dbdbdb;
  color: #323a46;
  cursor: pointer;
}

.c16 {
  background-clip: content-box;
  background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAwAAAAMCAIAAADZF8uwAAAAGUlEQVQYV2M4gwH+YwCGIasIUwhT25BVBADtzYNYrHvv4gAAAABJRU5ErkJggg==');
  border-radius: .2rem;
}

.c16 .c25,
.c16 .c20 {
  color: #fff;
  -webkit-transition: none;
  transition: none;
}

.c16 .c25[data-src-is-dark="true"],
.c16 .c20[data-src-is-dark="true"] {
  color: #2b2d31;
}

.c8 {
  border: solid 2px #fff;
  border-radius: 100%;
  box-shadow: 0 0 5px rgba(0,0,0,.5);
  box-sizing: border-box;
  position: absolute;
}

.c8[data-src-is-dark="true"] {
  border-color: #2b2d31;
}

.c9::after {
  border: solid 4px #fff;
  border-radius: 100%;
  content: "";
  left: calc(50% - 4px);
  position: absolute;
  top: calc(50% - 4px);
}

.c9[data-src-is-dark="true"]::after {
  border-color: #2b2d31;
}

.c4 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  overflow: hidden;
  position: relative;
  touch-action: none;
  cursor: pointer;
}

.c4,
.c4 div {
  border-radius: 0.5rem;
}

.c4 .src-color-picker-major-picker-bg {
  border-radius: calc(0.5rem + 1px);
}

.c4 div {
  height: 100%;
}

.c11 {
  background: linear-gradient(to bottom,#f00,#ff0,#0f0,#0ff,#00f,#f0f,#f00);
  box-shadow: 0 0 3px rgba(0,0,0,.3) inset;
}

<div
    class="c0 c1"
  >
    <section
      class="c2"
    >
      <div
        class="c3 c4"
        style="background: rgb(255, 0, 0);"
        value="rgb(255, 0, 0)"
      >
        <div
          class="src-color-picker-major-picker-bg"
        >
          <div
            class="c5"
          >
            <div
              class="c6"
            >
              <span
                class="c7 c8 c9"
                data-src-is-dark="false"
                style="left: 0px; top: 0px;"
                value="rgb(255, 0, 0)"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="c10 c4 c11"
        value="rgb(255, 0, 0)"
      >
        <span
          class="c12 c8"
          style="left: 0px; top: 0px;"
          value="rgb(255, 0, 0)"
        />
      </div>
      <div
        class="c10 c4"
        value="rgb(255, 0, 0)"
      >
        <div
          class="c13"
        >
          <div
            class="c14"
          >
            <span
              class="c12 c8"
              data-src-is-dark="false"
              style="left: 0px; top: 0px;"
              value="rgb(255, 0, 0)"
            />
          </div>
        </div>
      </div>
    </section>
    <hr
      class="c15"
    />
    <section
      class=" c16"
    >
      <label
        class="c17 c18"
      >
        <div
          class="c19"
        >
          <label
            class="c20 c21 c22 src-input-addon "
            data-src-is-dark="false"
            shape="rectangle"
            style="background: rgb(255, 0, 0);"
          >
            <svg
              class="c23 "
              viewBox="0 0 24 24"
            >
              <path
                d="M12 3c-4.97 0-9 4.03-9 9s4.03 9 9 9c.83 0 1.5-.67 1.5-1.5 0-.39-.15-.74-.39-1.01-.23-.26-.38-.61-.38-.99 0-.83.67-1.5 1.5-1.5H16c2.76 0 5-2.24 5-5 0-4.42-4.03-8-9-8zm-5.5 9c-.83 0-1.5-.67-1.5-1.5S5.67 9 6.5 9 8 9.67 8 10.5 7.33 12 6.5 12zm3-4C8.67 8 8 7.33 8 6.5S8.67 5 9.5 5s1.5.67 1.5 1.5S10.33 8 9.5 8zm5 0c-.83 0-1.5-.67-1.5-1.5S13.67 5 14.5 5s1.5.67 1.5 1.5S15.33 8 14.5 8zm3 4c-.83 0-1.5-.67-1.5-1.5S16.67 9 17.5 9s1.5.67 1.5 1.5-.67 1.5-1.5 1.5z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </label>
          <div
            class=" c24"
            shape="rectangle"
          >
            <input
              class="c25 c26 c27 src-input "
              data-src-is-dark="false"
              shape="rectangle"
              style="background: rgb(255, 0, 0);"
              type="text"
            />
          </div>
          <label
            class="c20 c21 c28 src-input-addon "
            data-src-is-dark="false"
            shape="rectangle"
            style="background: rgb(255, 0, 0);"
          >
            <svg
              class="c23 "
              viewBox="0 0 2048 1792"
            >
              <path
                d="M960 1408l336-384h-768l-336 384h768zm1013-1077q15 34 9.5 71.5t-30.5 65.5l-896 1024q-38 44-96 44h-768q-38 0-69.5-20.5t-47.5-54.5q-15-34-9.5-71.5t30.5-65.5l896-1024q38-44 96-44h768q38 0 69.5 20.5t47.5 54.5z"
              />
            </svg>
          </label>
        </div>
      </label>
    </section>
    <section
      class=""
    >
      <div
        class="c29"
      >
        <button
          class="c30 c31"
          shape="rectangle"
        >
          <span
            class="c32 src-button-content"
          >
            <svg
              class="c23 "
              viewBox="0 0 24 24"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M17 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V7l-4-4zm-5 16c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3zm3-10H5V5h10v4z"
              />
            </svg>
            <span>
              Save
            </span>
          </span>
        </button>
        <button
          class="c33 c31"
          shape="rectangle"
        >
          <span
            class="c32 src-button-content"
          >
            <svg
              class="c23 "
              viewBox="0 0 1792 1792"
            >
              <path
                d="M1490 1322q0 40-28 68l-136 136q-28 28-68 28t-68-28l-294-294-294 294q-28 28-68 28t-68-28l-136-136q-28-28-28-68t28-68l294-294-294-294q-28-28-28-68t28-68l136-136q28-28 68-28t68 28l294 294 294-294q28-28 68-28t68 28l136 136q28 28 28 68t-28 68l-294 294 294 294q28 28 28 68z"
              />
            </svg>
            <span>
              Cancel
            </span>
          </span>
        </button>
      </div>
    </section>
    <section
      class=""
    >
      <svg
        class="c23 c34"
        viewBox="0 0 1792 1792"
      >
        <path
          d="M1395 864q0 13-10 23l-466 466q-10 10-23 10t-23-10l-466-466q-10-10-10-23t10-23l50-50q10-10 23-10t23 10l393 393 393-393q10-10 23-10t23 10l50 50q10 10 10 23zm0-384q0 13-10 23l-466 466q-10 10-23 10t-23-10l-466-466q-10-10-10-23t10-23l50-50q10-10 23-10t23 10l393 393 393-393q10-10 23-10t23 10l50 50q10 10 10 23z"
        />
      </svg>
    </section>
  </div>
</DocumentFragment>
`;
